File: src/FlashcardParser/StudySetParser.ts
import { categoriesOf, Category } from "./CategoriesFinder";
import { parseCardLine } from "./SingleLineParser/Card";
import { countTabs } from "./LineTabCounter";
import {
    CATEGORIES,
    COMMENT_SYMBOL
} from './Symbols';
import {
    IStudySet,
    LineDescriptor
} from "./Types/Types";
import { parseCommandLine } from "./SingleLineParser/Command";
import { parseAcronymLine } from "./SingleLineParser/Alias";
import { parseResourceLine } from "./SingleLineParser/Resource";

export function parseStudyset(lines: string[]): IStudySet | null {
    const originalLines = lines;
    const categories = categoriesOf(lines);
    const categoriesAmount = categories.length;
    const studySet: IStudySet = {
        title: "",
        flashcards: [],
        resources: {},
        defaultResource: "",
        aliases: [],
        headers: [],
        studiedCards: 0,
        originalLines: [...originalLines],
        linesDescriptors: []
    };

    try {
        let lastCategoryIndex = 0;
        let currentCategory: Category = undefined;
        let isUnderComment: boolean = false;


        for (let i = 0; i < lines.length; i++) {
            /*
                Check if the current line is under a different category from
                old lines.
            */
            for (let j = lastCategoryIndex; j < categoriesAmount; j++) {
                const category = categories[j];
                const { line } = category;
                if (line > i) break;

                if (j != lastCategoryIndex) {
                    console.log(`[parser] Reached SECTION '${JSON.stringify(category)}'`);
                }
                currentCategory = category;
                lastCategoryIndex = j;
            }

            /*
                A line is considered a commet if it starts with a comment symbol
                or is tabbed under a commented line.
            */
            const lineTabs = countTabs(lines[i]);
            const trimmedLine = lines[i].trim();
            const isComment = trimmedLine.startsWith(COMMENT_SYMBOL);

            if (isComment || (lineTabs > 0 && isUnderComment)) {
                isUnderComment = true;
            } else {
                isUnderComment = false;
            }

            /*
                Creates the LineDescriptor object for this line and add it
                to the study set.
            */
            const lineDescriptor: LineDescriptor = {
                index: i,
                originalLine: lines[i],
                trimmedLine: trimmedLine,
                tabs: lineTabs,
                category: currentCategory,
                tabbedUnder: [],
                parent: null,
                isComment: isComment,
                subParts: []
            };
            studySet.linesDescriptors.push(lineDescriptor);
        }

        /*
            For each line, we need to know what lines are tabbed under it and
            who is the parent.
        */
       for (let k = 0; k < studySet.linesDescriptors.length; k++) {
            const lineDescriptor = studySet.linesDescriptors[k];
            const { tabs } = lineDescriptor;

            if (lineDescriptor.isComment) continue;
            if (tabs === 0) continue;
            
            for (let i = k - 1; i >= 0; i--) {
                const previousLd = studySet.linesDescriptors[i];
                if (previousLd.tabs == tabs) continue;
                if (previousLd.tabs < tabs) {
                    lineDescriptor.parent = previousLd;
                    previousLd.tabbedUnder.push(lineDescriptor);
                }
                break;
            }
        }

        /*
            Parse these lines. What the line does heavily depends
            on the context of this line, i.e. it's content and what there
            is around it. All these information are gathered in the
            LineContext object.
        */
        
        for (let lineDescriptor of studySet.linesDescriptors.filter(ld => !ld.isComment)) {
            const result = parseLine(lineDescriptor, studySet);

            if (result === false) {
                console.error(`[parser] Fail at line ${lineDescriptor.index}:` +
                    `"${lineDescriptor.originalLine}"`);
                return null;
            }
        }


        return studySet;
    } catch (error) {
        console.error(`[parser] ERROR: ${error}`);
        return null;
    }
}



function parseLine(
    lineDescriptor: LineDescriptor,
    studySet: IStudySet
): boolean {
    const { trimmedLine, category, index } = lineDescriptor;
    const { name: categoryName, line: categoryLine } = category;
    
    if (trimmedLine.length == 0) return true;
    if (index == categoryLine) return true;
    if (lineDescriptor.isComment) return true;

    if (lineDescriptor.tabs > 0) {
        return parseCommandLine(lineDescriptor, studySet);
    }

    switch (categoryName) {
        case CATEGORIES.title:
            console.log(`[parser] Reading title '${trimmedLine}'`);
        
            if (studySet.title) {
                console.error(`[parser] Multiple title lines found: ` +
                    `"${studySet.title}" and "${trimmedLine}"`);
                return false;
            }
            studySet.title = trimmedLine;
            break;
        case CATEGORIES.resources:
            console.log(`[parser] Reading resource '${trimmedLine}'`);
            return parseResourceLine(lineDescriptor, studySet);
        case CATEGORIES.cards:
            console.log(`[parser] Read card at line ${index}`);
            return parseCardLine(lineDescriptor, studySet);
        case CATEGORIES.acronyms:
            console.log(`[parser] Reading acronym '${trimmedLine}'`);
            return parseAcronymLine(lineDescriptor, studySet);
        default:
            console.log('[parser] Unrecognized category: '+ categoryName);
            return false;
    }

    return true;
}

File: service/flashcard_parser/studyset_parser.py
from typing import List

from .categories_finder import categories_of, Category
from .single_line_parser.card import parse_card_line
from .line_tab_counter import count_tabs
from .symbols import CATEGORIES, COMMENT_SYMBOL
from .types import StudySet, LineDescriptor
from .single_line_parser.command import parse_command_line
from .single_line_parser.alias import parse_acronym_line
from .single_line_parser.resource import parse_resource_line


def parse_studyset(lines: List[str]) -> StudySet | None:
    original_lines = list(lines)
    categories = categories_of(lines)
    study_set = StudySet(original_lines=original_lines)
    try:
        last_category_index = 0
        current_category: Category | None = None
        is_under_comment = False
        for i, line in enumerate(lines):
            for j in range(last_category_index, len(categories)):
                category = categories[j]
                if category.line > i:
                    break
                current_category = category
                last_category_index = j
            line_tabs = count_tabs(line)
            trimmed_line = line.strip()
            is_comment = trimmed_line.startswith(COMMENT_SYMBOL)
            if is_comment or (line_tabs > 0 and is_under_comment):
                is_under_comment = True
            else:
                is_under_comment = False
            ld = LineDescriptor(
                index=i,
                original_line=line,
                trimmed_line=trimmed_line,
                tabs=line_tabs,
                category=current_category,
                tabbed_under=[],
                parent=None,
                is_comment=is_comment,
                sub_parts=[],
            )
            study_set.lines_descriptors.append(ld)
        for k, ld in enumerate(study_set.lines_descriptors):
            if ld.is_comment or ld.tabs == 0:
                continue
            for i in range(k - 1, -1, -1):
                prev = study_set.lines_descriptors[i]
                if prev.tabs >= ld.tabs:
                    continue
                if prev.tabs < ld.tabs:
                    ld.parent = prev
                    prev.tabbed_under.append(ld)
                break
        for ld in [x for x in study_set.lines_descriptors if not x.is_comment]:
            result = _parse_line(ld, study_set)
            if result is False:
                print(f"[parser] Fail at line {ld.index}: '{ld.original_line}'")
                return None
        return study_set
    except Exception as e:
        print(f"[parser] ERROR: {e}")
        return None


def _parse_line(line_descriptor: LineDescriptor, study_set: StudySet) -> bool:
    trimmed_line = line_descriptor.trimmed_line
    category = line_descriptor.category
    category_name = category.name if category else None
    category_line = category.line if category else None
    index = line_descriptor.index
    if len(trimmed_line) == 0:
        return True
    if category_line == index:
        return True
    if line_descriptor.is_comment:
        return True
    if line_descriptor.tabs > 0:
        return parse_command_line(line_descriptor, study_set)
    if category_name == CATEGORIES["title"]:
        if study_set.title:
            print(f"[parser] Multiple title lines found: '{study_set.title}' and '{trimmed_line}'")
            return False
        study_set.title = trimmed_line
    elif category_name == CATEGORIES["resources"]:
        return parse_resource_line(line_descriptor, study_set)
    elif category_name == CATEGORIES["cards"]:
        return parse_card_line(line_descriptor, study_set)
    elif category_name == CATEGORIES["acronyms"]:
        return parse_acronym_line(line_descriptor, study_set)
    else:
        print(f"[parser] Unrecognized category: {category_name}")
        return False
    return True
